firestore/docref.go

```diff
@@ -638,6 +638,28 @@ func (s sentinel) String() string {
 	}
 }
 
+// VectorType represpresents a vector
+type VectorType interface {
+	isVectorType()
+	toProtoValue() (*pb.Value, bool, error)
+}
+
+// Vector represents a vector in the form of a float64 array
+type Vector []float64
+
+func (_ Vector) isVectorType() {}
+
+func (vector Vector) toProtoValue() (*pb.Value, bool, error) {
```

Comment:
This can't fail or have transforms, so it can just return *pb.Value. Also we can avoid the reflection overhead of mapToProtoValue by just writing out the protos, like this:
```
// toProtoValue returns a Firestore [pb.Value] representing the Vector.                                            
func (v Vector) toProtoValue() *pb.Value {
        vals := make([]*pb.Value, len(v))            
        for i, val := range v {
                vals[i] = floatToProtoValue(float64(val))                                                          
        }
        return &pb.Value{          
                ValueType: &pb.Value_MapValue{
                        MapValue: &pb.MapValue{                                                                    
                                Fields: map[string]*pb.Value{                                                      
                                        typeKey: stringToProtoValue(vectorValue),
                                        "value": &pb.Value{
                                                ValueType: &pb.Value_ArrayValue{
                                                        ArrayValue: &pb.ArrayValue{Values: vals},
                                                },
                                        },
                                },
                        },
                },
        }
}
```

Reply:
```
func stringToProtoValue(s string) *pb.Value {
        return &pb.Value{ValueType: &pb.Value_StringValue{StringValue: s}}
}

func floatToProtoValue(f float64) *pb.Value {
        return &pb.Value{ValueType: &pb.Value_DoubleValue{DoubleValue: f}}
}
```

Reply:
In my code I forgot the `v == nil` case. Please add.

Reply:
Refactored